CompilerSource/parsing/pretty_printer.cpp

```diff
@@ -0,0 +1,733 @@
+/** Copyright (C) 2024 Fares Atef
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <JDI/src/System/builtins.h>
+#include "ast.h"
+
+using namespace enigma::parsing;
+
+bool AST::Visitor::VisitIdentifierAccess(IdentifierAccess &node) {
+  print(std::string(node.name.content));
+  return true;
+}
+
+bool AST::Visitor::VisitLiteral(Literal &node) {
+  std::string value = std::get<std::string>(node.value.value);
+  if (value == "'\n'") {
+    print("'\\n'");
+  } else if (value == "'\t'") {
+    print("'\\t'");
+  } else if (value == "'\v'") {
+    print("'\\v'");
+  } else if (value == "'\b'") {
+    print("'\\b'");
+  } else if (value == "'\r'") {
+    print("'\\r'");
+  } else if (value == "'\f'") {
+    print("'\\f'");
+  } else if (value == "'\a'") {
+    print("'\\a'");
+  } else if (value == "'\\'") {
+    print("'\\\'");
+  } else if (value == "'\''") {
+    print("'\\''");
+  } else if (value == "'\"'") {
+    print("'\\\"'");
+  } else if (value == "'\?'") {
+    print("'\\?'");
+  } else {
+    print(value);
+  }
+  return true;
+}
+
+bool AST::Visitor::VisitParenthetical(Parenthetical &node) {
+  bool printed = false;
+  print("(");
+
+  printed = Visit(node.expression);
```

Comment:
nit: Since you only really assign this once in most methods, it's probably better to just inline it (`if (!Visit(subnode)) return false;`).

Reply:
Nice idea, it is cleaner now. I also introduced a `VISIT_AND_CHECK` macro.